sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+        return lit.arg
+
+    if _type == 'AND':
+        return -2, lit.args
+
+    if _type == 'OR':
+        return -1, lit.args
+
+
+class OR(object):
+    """
+    A low-level implementation for Or
+    """
+    def __init__(self, *args):
+        self._args = args
+
+    @property
+    def args(self):
+        return sorted(self._args, key=literal_sort_key)
+
+    def __invert__(self):
+        return AND(*[~arg for arg in self._args])
+
+    def __hash__(self):
+        return hash((type(self).__name__, ) + tuple(self.args))
+
+    def __eq__(self, other):
+        return self.args == other.args
+
+    def __str__(self):
+        s = '(' + ' | '.join([str(arg) for arg in self.args]) + ')'
+        return s
+
+    __repr__ = __str__
+
+
+class AND(object):
+    """
+    A low-level implementation for And
+    """
+    def __init__(self, *args):
+        self._args = args
+
+    def __invert__(self):
+        return OR(*[~arg for arg in self._args])
+
+    @property
+    def args(self):
+        return sorted(self._args, key=literal_sort_key)
+
+    def __hash__(self):
+        return hash((type(self).__name__, ) + tuple(self.args))
+
+    def __eq__(self, other):
+        return self.args == other.args
+
+    def __str__(self):
+        s = '('+' & '.join([str(arg) for arg in self.args])+')'
+        return s
+
+    __repr__ = __str__
+
+
+def to_NNF(expr):
+    """
+    Generates the Negation Normal Form of any boolean expression in terms
+    of AND , OR and Literal objects.
+    """
+    if not isinstance(expr, BooleanFunction):
+        return Literal(expr)
+
+    klass = type(expr).__name__
+    if klass == 'Not':
+        arg = expr.args[0]
+        tmp = to_NNF(arg)  # Strategy: negate the NNF of expr
+        return ~tmp
+
+    if klass == 'Or':
+        return OR(*[to_NNF(x) for x in Or.make_args(expr)])
+
+    if klass == 'And':
+        return AND(*[to_NNF(x) for x in And.make_args(expr)])
+
+    if klass == 'Nand':
+        tmp = AND(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Nor':
+        tmp = OR(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Xor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
+                clause = [~to_NNF(s) if s in neg else to_NNF(s)
+                          for s in expr.args]
+                cnfs.append(OR(*clause))
+        return AND(*cnfs)
+
+    if klass == 'Xnor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
+                clause = [~to_NNF(s) if s in neg else to_NNF(s)
+                          for s in expr.args]
+                cnfs.append(OR(*clause))
+        return ~AND(*cnfs)
+
+    if klass == 'Implies':
+        L, R = map(to_NNF, expr.args)
+        return OR(~L, R)
+
+    if klass == 'Equivalent':
+        cnfs = []
+        for a, b in zip(expr.args, expr.args[1:]):
```

Comment:
You can use zip_longest (from itertools)

```
>>> a = range(4)
>>> from itertools import zip_longest
>>> list(zip_longest(a, a[1:], fillvalue=a[0]))
[(0, 1), (1, 2), (2, 3), (3, 0)]
```

Reply:
`zip_longest` is not present in Python2, instead the function is `izip_longest`.

Reply:
This is handled in sympy.compatibility so import it from there